modal/app.py

```diff
@@ -682,55 +677,82 @@ def cls(
-            secrets=secrets,
-            gpu=gpu,
-            serialized=serialized,
-            mounts=mounts,
-            network_file_systems=network_file_systems,
-            allow_cross_region_volumes=allow_cross_region_volumes,
-            volumes=volumes,
-            cpu=cpu,
-            memory=memory,
-            ephemeral_disk=ephemeral_disk,
-            proxy=proxy,
-            retries=retries,
-            concurrency_limit=concurrency_limit,
-            allow_concurrent_inputs=allow_concurrent_inputs,
-            container_idle_timeout=container_idle_timeout,
-            timeout=timeout,
-            interactive=interactive,
-            keep_warm=keep_warm,
-            cloud=cloud,
-            region=region,
-            enable_memory_snapshot=enable_memory_snapshot,
-            checkpointing_enabled=checkpointing_enabled,
-            block_network=block_network,
-            _allow_background_volume_commits=_allow_background_volume_commits,
-            max_inputs=max_inputs,
-            _experimental_boost=_experimental_boost,
-            _experimental_scheduler=_experimental_scheduler,
-            _experimental_scheduler_placement=_experimental_scheduler_placement,
-        )
+        if isinstance(_warn_parentheses_missing, _Image):
+            # Handle edge case where maybe (?) some users passed image as a positional arg
+            raise InvalidError("`image` needs to be a keyword argument: `@app.function(image=image)`.")
+        if _warn_parentheses_missing:
+            raise InvalidError("Did you forget parentheses? Suggestion: `@app.function()`.")
+
+        if interactive:
+            deprecation_error(
+                (2024, 5, 1), "interactive=True has been deprecated. Set MODAL_INTERACTIVE_FUNCTIONS=1 instead."
+            )
+
+        if image is None:
+            image = self._get_default_image()
+
+        secrets = [*self._secrets, *secrets]
 
         def wrapper(user_cls: CLS_T) -> _Cls:
-            cls: _Cls = _Cls.from_local(user_cls, self, decorator)
+            nonlocal keep_warm
+
+            # Check if the decorated object is a class
+            if not inspect.isclass(user_cls):
+                raise TypeError("The @app.cls decorator must be used on a class.")
+
+            info = FunctionInfo(None, serialized=serialized, cls=user_cls)
+            webhook_config = None
+
+            scheduler_placement: Optional[SchedulerPlacement] = _experimental_scheduler_placement
+            if region:
+                if scheduler_placement:
+                    raise InvalidError("`region` and `_experimental_scheduler_placement` cannot be used together")
+                scheduler_placement = SchedulerPlacement(region=region)
+
+            cls_func = _Function.from_args(
+                info,
+                app=self,
+                image=image,
+                secret=secret,
+                secrets=secrets,
+                is_generator=False,  # TODO: weird?
```

Comment:
I don't think we support that and if we did, `__call__` would still just be a method on the "class service function", which in itself doesn't have a valid value for `is_generator` since it isn't a callable in that way.

I made this a bit clearer now by moving `is_generator` and `webhook_config` (both of which are only valid in the context of a callable) to the end of the from_args call, with an explaining comment